src/tls/certificate_manager.rs

```diff
@@ -0,0 +1,234 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashMap;
+use std::fs::{ self };
+use std::io::{ Error, ErrorKind };
+use std::path::PathBuf;
+use std::sync::{ Arc, Mutex, RwLock };
+
+use tls::certificate_record::CertificateRecord;
+use tls::ssl_context::SslContextProvider;
+
+#[derive(Clone)]
+pub struct CertificateManager {
+    ssl_hosts: Arc<RwLock<HashMap<String, CertificateRecord>>>,
+
+    // Observer
+    context_provider: Option<Arc<Mutex<Box<SslContextProvider>>>>
+}
+
+fn create_records_from_directory(path: &PathBuf) -> Result<HashMap<String, CertificateRecord>, Error> {
+    let mut records = HashMap::new();
+
+    if try!(fs::metadata(path)).is_dir() {
+        for entry in try!(fs::read_dir(path)) {
+            let entry = try!(entry);
+
+            let hostname = entry.file_name().into_string().unwrap();
+            info!("Using certificate for host {}", hostname);
+
+            let mut host_path = path.clone();
+            host_path.push(hostname.clone());
+
+            let mut cert_path = host_path.clone();
+            cert_path.push("crt.pem");
+
+            let mut private_key_file = host_path.clone();
+            private_key_file.push("private_key.pem");
+
+            records.insert(hostname.clone(), CertificateRecord {
+                hostname: hostname,
+                private_key_file: private_key_file,
+                cert_file: cert_path
+            });
+        }
+
+        info!("Loaded certificates from directory: {:?}", path);
+
+        Ok(records)
+    } else {
+        Err(Error::new(
+            ErrorKind::InvalidInput,
+            "The configured SSL certificate directory is not recognised as a directory."
+        ))
+    }
+}
+
+impl CertificateManager {
+    pub fn new() -> CertificateManager {
+        CertificateManager {
+            ssl_hosts: Arc::new(RwLock::new(HashMap::new())),
+            context_provider: None,
+        }
+    }
+
+    pub fn set_context_provider(&mut self, context_provider: Arc<Mutex<Box<SslContextProvider>>>) {
+        if self.context_provider.is_none() {
+           self.context_provider = Some(context_provider);
+           self.notify_provider();
+        }
```

Comment:
Should we throw (or log an error) if the context provider is set more than once?